pkg/provider/naming.go

```diff
@@ -0,0 +1,44 @@
+package provider
+
+import (
+	"fmt"
+	"math/rand"
+
+	"github.com/pulumi/pulumi/pkg/tokens"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+var renameBlacklist = sets.NewString("v1/Namespace")
+
+// assignName generates a name for an object if it doesn't already have one. Uses DNS-1123-compliant
+// characters.
+func assignName(obj *unstructured.Unstructured, base tokens.QName) {
+	if !renameBlacklist.Has(obj.GroupVersionKind().String()) {
+		replaceNameIfUnnamed(obj, fmt.Sprintf("%s-%s", base, randString(8)))
+	} else {
+		replaceNameIfUnnamed(obj, string(base))
+	}
+}
+
+// replaceNameIfUnnamed will replace an object's name if it doesn't already have one.
+func replaceNameIfUnnamed(obj *unstructured.Unstructured, name string) {
+	if obj.GetName() == "" {
+		meta := obj.Object["metadata"]
+		if meta == nil {
+			obj.Object["metadata"] = map[string]interface{}{"name": name}
+		} else {
+			meta.(map[string]interface{})["name"] = name
+		}
+	}
+}
+
+func randString(n int) string {
+	var alphabet = []rune("abcdefghijklmnopqrstuvwxyz0123456789")
+
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = alphabet[rand.Intn(len(alphabet))]
```

Comment:
`rand.Intn` is deterministic by default (https://github.com/golang/go/issues/4509) - since I don't think we seed rand anywhere we should either 1) seed it explicitly on startup or 2) use `crypto/rand` so that we don't get in bad situations with name collisions?

Reply:
Ah, OK, looks like we are seeding:

https://github.com/pulumi/pulumi-kubernetes/blob/8c0b67e07348696c273cfb109d1997c22a248c44/pkg/watcher/watcher.go#L188

Never mind, then.

Reply:
We should still be explicit about the seeding in this package if it requires it.